cou/steps/nova_cloud_controller.py

```diff
@@ -63,6 +64,46 @@ async def archive(model: Model, *, batch_size: int) -> None:
         logger.debug("Potentially more data to archive...")
 
 
+async def purge(model: Model, before: Optional[str]) -> None:
+    """Purge data on a nova-cloud-controller unit.
+
+    The purge-data action delete rows from shadow tables.
+    :param model: juju model to work with
+    :type model: Model
+    :param before: specifying before will delete data from all shadow tables that is older than the data provided.
+         Date strings may be fuzzy, such as Oct 21 2015
+    :raises COUException: if action returned unexpected output or failed
+    """
+    action_params = {}
+    if before is not None:
+        action_params = {"before": before}
+
+    unit_name: str = await _get_nova_cloud_controller_unit_name(model)
+    action = await model.run_action(
```

Comment:
What if there's no such action? To provide a better user experience, we should check if the action exists or not before running this action (just like what was done for getting config options) not just crashing COU. Warning message or using try-except maybe good enough

Reply:
https://github.com/canonical/charmed-openstack-upgrader/pull/444#discussion_r1645328802

Reply:
Same as https://github.com/canonical/charmed-openstack-upgrader/pull/444#discussion_r1645115931

resolved.